tests/zkevm/test_worst_bytecode.py

```diff
@@ -65,10 +63,45 @@ def test_worst_bytecode_single_opcode(
     The test is performed in the last block of the test, and the entire block gas limit is
     consumed by repeated opcode executions.
     """
-    # We use 100G gas limit to be able to deploy a large number of contracts in a single block,
-    # avoiding bloating the number of preparing blocks in the test.
-    env = Environment(gas_limit=100_000_000_000)
-    attack_gas_limit = Environment().gas_limit
+    # The attack gas limit is the gas limit which the target tx will use
+    # The test will scale the block gas limit to setup the contracts accordingly to be
+    # able to pay for the contract deposit. This has to take into account the 200 gas per byte,
+    # but also the quadratic memory expansion costs which have to be paid each time the
+    # memory is being setup
+    attack_gas_limit = 100_000_000
+    max_contract_size = fork.max_code_size() or 24 * 1024  # Uses Spurious Dragon limit by default
+
+    gas_costs = fork.gas_costs()
+
+    # Calculate the absolute minimum gas costs to deploy the contract
+    # This does not take into account setting up the actual memory (using KECCAK256 and XOR)
+    # so the actual costs of deploying the contract is higher
+    memory_expansion_gas_calculator = fork.memory_expansion_gas_calculator()
+    memory_gas_minimum = memory_expansion_gas_calculator(new_bytes=len(bytes(max_contract_size)))
+    code_deposit_gas_minimum = (
+        fork.gas_costs().G_CODE_DEPOSIT_BYTE * max_contract_size + memory_gas_minimum
+    )
+
+    intrinsic_gas_cost_calc = fork.transaction_intrinsic_cost_calculator()
+    # Calculate the loop cost of the attacker to query one address
+    loop_cost = (
+        gas_costs.G_KECCAK_256  # KECCAK static cost
+        + math.ceil(85 / 32) * gas_costs.G_KECCAK_256_WORD  # KECCAK dynamic cost for CREATE2
+        + gas_costs.G_VERY_LOW * 3  # ~MSTOREs+ADDs
+        + gas_costs.G_COLD_ACCOUNT_ACCESS  # Opcode cost
+        + 30  # ~Gluing opcodes
+    )
+    # Calculate the number of contracts to be targeted
+    num_contracts = (
+        # Base available gas = GAS_LIMIT - intrinsic - (out of loop MSTOREs)
+        attack_gas_limit - intrinsic_gas_cost_calc() - gas_costs.G_VERY_LOW * 4
+    ) // loop_cost
+
+    # Set the block gas limit to a relative high value to ensure the code deposit tx
+    # fits in the block (there is enough gas available in the block to execute this)
+    # TODO: verify this, I think we can just do `code_deposit_gas_minimum * 2 * num_contracts`?
```

Comment:
> TODO: verify this, I think we can just do `code_deposit_gas_minimum * 2 * num_contracts`?

I think so, but I would suggest to debug this test as is and see that there's a big buffer. This is why I prefered originally to set up simply a 100Ggas... trying to do some math gymnastics to later do `2*` sounds like overcomplicating the setup.

We can even put 1 trillion gas or something and call it a day to cover an attack scenario that is much bigger than today. (Note that if is 1 trillion or tighter to the actual need, it doesn't make filling faster or similar).

Reply:
Oh right, good point. No idea why I did not take this super high constant but instead lowered it :sweat_smile:

I thought of adding this as assertion, but this is not necessary, since post state has the check that the nonces of the predeployed contracts are 1 (so contract is deployed) :smile: :+1: